crates/accelerate/src/two_qubit_decompose.rs

```diff
@@ -0,0 +1,179 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// In numpy matrices real and imaginary components are adjacent:
+//   np.array([1,2,3], dtype='complex').view('float64')
+//   array([1., 0., 2., 0., 3., 0.])
+// The matrix faer::Mat<c64> has this layout.
+// faer::Mat<num_complex::Complex<f64>> instead stores a matrix
+// of real components and one of imaginary components.
+// In order to avoid copying we want to use `MatRef<c64>` or `MatMut<c64>`.
+
+use num_complex::Complex;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+use std::f64::consts::PI;
+
+use faer::IntoFaerComplex;
+use faer::{mat, prelude::*, scale, Mat, MatRef};
+use faer_core::{c64, ComplexField};
+use numpy::PyReadonlyArray2;
+
+use crate::utils;
+
+// FIXME: zero and one exist but I cant find the right incantation
+const C0: c64 = c64 { re: 0.0, im: 0.0 };
+const C1: c64 = c64 { re: 1.0, im: 0.0 };
+const C1IM: c64 = c64 { re: 0.0, im: 1.0 };
+
+// FIXME: Find a way to compute these matrices at compile time.
+fn transform_from_magic_basis(unitary: Mat<c64>) -> Mat<c64> {
+    let _b_nonnormalized: Mat<c64> = mat![
+        [C1, C1IM, C0, C0],
+        [C0, C0, C1IM, C1],
+        [C0, C0, C1IM, -C1],
+        [C1, -C1IM, C0, C0]
+    ];
+    let _b_nonnormalized_dagger = scale(c64 { re: 0.5, im: 0.0 }) * _b_nonnormalized.adjoint();
+    _b_nonnormalized_dagger * unitary * _b_nonnormalized
+}
+
+// faer::c64 and num_complex::Complex<f64> are both structs
+// holding two f64's. But several functions are not defined for
+// c64. So we implement them here. These things should be contribute
+// upstream.
+
+pub trait PowF {
+    fn powf(self, pow: f64) -> c64;
+}
+
+impl PowF for c64 {
+    fn powf(self, pow: f64) -> c64 {
+        c64::from(self.to_num_complex().powf(pow))
+    }
+}
+
+pub trait Arg {
+    fn arg(self) -> f64;
+}
+
+impl Arg for c64 {
+    fn arg(self) -> f64 {
+        self.to_num_complex().arg()
+    }
+}
+
+fn __weyl_coordinates(unitary: MatRef<c64>) -> (f64, f64, f64) {
```

Comment:
I'd personally use an array return type here because it's a bit more compact (so we don't need to write `f64` 3 times):

```suggestion
fn __weyl_coordinates(unitary: MatRef<c64>) -> [f64; 3] {
```

Performance wise I don't think it'll make any difference